### PR TITLE
[NPU][Perf] Run MiMo-V2.5 MoE router in bf16 and drop redundant topk_ids casts

### PR DESCRIPTION
**Pull request by iridiumine**
## Motivation

The MiMo-V2.5 MoE router kept its gate weight in fp32: every MoE layer cast `hidden_states` bf16→fp32 for the router matmul, then `fused_topk_npu` cast the logits back to fp32 for `npu_moe_gating_top_k` — two Cast kernels plus a non-Cube fp32 matmul per layer (57 layers). Separately, the DeepEP low-latency dispatcher converted `topk_ids` to int64 while the NPU DeepEP runtime takes int32 — an int32→int64→int32 round trip costing 2–3 more Cast kernels per layer.

## Modifications

- Keep `MoEGate.weight` / `e_score_correction_bias` in bf16 **on NPU**: the router matmul runs natively on Cube units and `npu_moe_gating_top_k` consumes bf16 logits directly. GPU keeps the historical fp32 gate.
- `fused_topk_npu`: dtype-alignment guard instead of unconditional fp32 casts — nothing is cast when logits and bias share a dtype; an fp32 bias promotes the logits to fp32 (downcasting the bias to bf16 reorders top-k routing, see the GLM-5.2 note in `deepseek_v2.py`). No bias → historical fp32 scoring, so sigmoid / `num_token_non_padded` models are unaffected.
- Skip `topk_ids.to(torch.int64)` in the DeepEP LL dispatcher on NPU; GPU keeps int64 (its native requirement).
- `topk_weights` stay fp32: `MoeLowLatencyCombineV2` requires DT_FLOAT expert scales — a hard requirement, not a leftover. The modelopt_fp4 + flashinfer_trtllm bias branch also stays fp32.

Other models: Qwen3-MoE-style (softmax, no bias) takes a different branch — unaffected; DeepSeek-family ends up fp32 either way — identical numerics; GPU behavior unchanged everywhere.

## Accuracy Tests

```
echo performance | tee /sys/devices/system/cpu/cpu*/cpufreq/scaling_governor
sysctl -w vm.swappiness=0
sysctl -w kernel.numa_balancing=0
sysctl -w kernel.sched_migration_cost_ns=50000
export SGLANG_SET_CPU_AFFINITY=1

source /usr/local/Ascend/ascend-toolkit/set_env.sh
source /usr/local/Ascend/nnal/atb/set_env.sh

export HCCL_BUFFSIZE=300
export HCCL_OP_EXPANSION_MODE=AIV
export PYTORCH_NPU_ALLOC_CONF=expandable_segments:True
export STREAMS_PER_DEVICE=32
export SGLANG_DISAGGREGATION_BOOTSTRAP_TIMEOUT=600
export SGLANG_ENABLE_SPEC_V2=1
export SGLANG_ENABLE_OVERLAP_PLAN_STREAM=1
export ASCEND_USE_FIA=1
export SGLANG_DEEPEP_NUM_MAX_DISPATCH_TOKENS_PER_RANK=32
export DEEPEP_HCCL_BUFFSIZE=2500
export DEEPEP_NORMAL_LONG_SEQ_ROUND=20
export DEEPEP_NORMAL_LONG_SEQ_PER_ROUND_TOKENS=4096
export HCCL_SOCKET_IFNAME=lo
export GLOO_SOCKET_IFNAME=lo
export HCCL_HOST_SOCKET_PORT_RANGE=auto
export SGLANG_ALLOW_OVERWRITE_LONGER_CONTEXT_LEN=1

sglang serve \
        --model-path $MODEL_PATH \
        --served-model-name $MODEL_PATH \
        --trust-remote-code \
        --attention-backend ascend \
        --device npu \
        --mem-fraction-static 0.92 \
        --tp-size 8 --nnodes 1 --node-rank 0 \
        --chunked-prefill-size 8192 --max-total-tokens 600000 \
        --max-running-requests 32 \
        --moe-a2a-backend deepep --deepep-mode auto \
        --cuda-graph-bs-decode 1 2 4 8 16 \
        --speculative-algorithm DFLASH \
        --speculative-draft-model-path $DFLASH_PATH \
        --speculative-num-draft-tokens 8 \
        --dp-size 2 --enable-dp-attention --enable-dp-lm-head \
```

```
DeprecationWarning: sglang.test.few_shot_gsm8k is deprecated. Use sglang.test.run_eval with eval_name='gsm8k' instead.
  metrics = run_eval(args)
100%|█████████████████████████████████████████████████████████████████████████████████| 200/200 [00:24<00:00,  8.24it/s]
Accuracy: 0.980
Invalid: 0.010
Latency: 24.645 s
Output throughput: 942.426 token/s
metrics={'accuracy': np.float64(0.98), 'invalid': np.float64(0.01), 'latency': 24.644899220205843, 'output_throughput': 942.4262518776089}
metrics['accuracy']=np.float64(0.98)
```

## Speed Tests and Profiling

Profiling (per MoE layer, decode, DeepEP LL, EP6):

| Kernel | Before | After |
|---|---|---|
| Router MatMulV3 | 13.84 us | 9.12 us (**-34%**) |
| Pre-router Cast (bf16→fp32) | 2.63 us | eliminated |
| Topk-related Casts | 4 casts / 8.67 us | 1 cast / 2.50 us |

The single remaining cast is the hard-required bf16→fp32 `topk_weights` conversion (CombineV2 DT_FLOAT).

Benchmark (16k-input prefill-heavy workload, 128 prompts, request rate 0.4):

| Metric | Before | After |
|---|---|---|
| Mean TTFT | 2308.5 ms | 1866.78 ms (**-19.1%**) |
| P90 TTFT | 4274.1 ms | 3033.30 ms (**-29.0%**) |
| P99 TTFT | 7061.9 ms | 4062.82 ms (**-42.5%**) |

Decode-side gains are minor as expected — the router is a small share of the decode step.

## Checklist

- [x] Format your code according to the [Format code with pre-commit](https://docs.sglang.io/developer_guide/contribution_guide.html#format-code-with-pre-commit).
- [ ] Add unit tests according to the [Run and add unit tests](https://docs.sglang.io/developer_guide/contribution_guide.html#run-and-add-unit-tests).
- [ ] Update documentation according to [Write documentations](https://docs.sglang.io/developer_guide/contribution_guide.html#write-documentations).
- [x] Provide accuracy and speed benchmark results according to [Test the accuracy](https://docs.sglang.io/developer_guide/contribution_guide.html#test-the-accuracy) and [Benchmark the speed](https://docs.sglang.io/developer_guide/contribution_guide.html#benchmark-the-speed).
- [x] Follow the SGLang code style [guidance](https://docs.sglang.io/developer_guide/contribution_guide.html#code-style-guidance).

## Review and Merge Process

1. Ping Merge Oncalls to start the process. See the [PR Merge Process](https://github.com/sgl-project/sglang/blob/main/.github/MAINTAINER.md#pull-request-merge-process).
2. Get approvals from [CODEOWNERS](https://github.com/sgl-project/sglang/blob/main/.github/CODEOWNERS) and other reviewers.
3. Trigger CI tests with [comments](https://docs.sglang.io/developer_guide/contribution_guide.html#how-to-trigger-ci-tests) or contact authorized users to do so.
   - Common commands include `/tag-and-rerun-ci`, `/tag-run-ci-label`, `/rerun-failed-ci`
4. After green CI and required approvals, ask Merge Oncalls or people with Write permission to merge the PR.

---
### CI States

Latest PR Test (Base): :x: [Run #35710012740](https://github.com/sgl-project/sglang/actions/runs/35710012740)
Latest PR Test (Extra): :x: [Run #35710012227](https://github.com/sgl-project/sglang/actions/runs/35710012227)
Latest PR Test (AMD ROCm 10): :x: [Run #35710012477](https://github.com/sgl-project/sglang/actions/runs/35710012477)

### CODE REVIEW

File: python/sglang/srt/hardware_backend/npu/moe/topk.py [hunk 1]
```diff
@@ -72,14 +72,23 @@ def fused_topk_npu(
...
+            # and only align a bf16 bias up to the logits dtype.
+            if correction_bias.dtype != router_logits.dtype:
+                if correction_bias.dtype == torch.float32:
+                    router_logits = router_logits.to(torch.float32)
+                else:
+                    correction_bias = correction_bias.to(router_logits.dtype)
+        else:
+            # No bias: keep the historical fp32 scoring.
+            router_logits = router_logits.to(torch.float32)
         topk_weights, topk_ids, _ = torch.ops.npu.npu_moe_gating_top_k(
```
Comment (iridiumine) [inline python/sglang/srt/hardware_backend/npu/moe/topk.py:88]: bf16 or fp32

Comment (iridiumine) [reply python/sglang/srt/hardware_backend/npu/moe/topk.py:88]: x (Tensor): Required. Input to be processed. This parameter must be a 2D tensor. The data type can be float16, bfloat16, or float32. The data layout must be ND. Non-contiguous tensors are supported. The size of the last dimension (the expert count) must be less than or equal to 2048.
https://www.hiascend.com/document/detail/en/Pytorch/2600/apiref/torchnpuCustomapi/docs/en/custom_APIs/torch_npu/torch_npu-npu_moe_gating_top_k.md
